File: generated_code/GeminiPro/Java/075_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * The splay tree is a self-balancing binary search tree with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the splay tree.
     * It is kept private to encapsulate the tree's internal structure.
     */
    private class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Default constructor initializes an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Core Splay Tree Operations: Rotations and Splaying ---

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The splay operation. It brings the node with the given key to the root of the tree.
     * If the key is not present, the last accessed node (the one that would be the parent
     * of the key if it were present) is brought to the root.
     * This is a recursive, bottom-up implementation.
     *
     * @param node The root of the current subtree.
     * @param key The key to splay.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        // Base cases: if the node is null or the key is at the root
        if (node == null || node.key == key) {
            return node;
        }

        // Key is in the left subtree
        if (key < node.key) {
            // If the left child is null, the key is not in the tree
            if (node.left == null) {
                return node;
            }
            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                node.left.left = splay(node.left.left, key);
                node = rightRotate(node);
            }
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                node.left.right = splay(node.left.right, key);
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Perform the second rotation (or the first if it's a Zig case)
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key is in the right subtree
        else {
            // If the right child is null, the key is not in the tree
            if (node.right == null) {
                return node;
            }
            // Zig-Zag (Right-Left)
            if (key < node.right.key) {
                node.right.left = splay(node.right.left, key);
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > node.right.key) {
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            }
            // Perform the second rotation (or the first if it's a Zig case)
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    // --- Public API for the Integer Set ---

    /**
     * Searches for a key in the tree. If found, the node containing the key is
     * splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a new key into the splay tree. Duplicates are not allowed.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // Handle the case of an empty tree
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. The closest node becomes the root.
        root = splay(root, key);

        // If the key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        // Create the new node
        Node newNode = new Node(key);

        // The old root becomes a child of the new node
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        // The new node is the new root
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete to the root
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        // We need to merge its left and right subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree
            root = rightSubtree;
        } else {
            // Find the maximum element in the left subtree by splaying for the key
            // (which is guaranteed to be the largest). This brings the max element
            // to the root of the left subtree.
            leftSubtree = splay(leftSubtree, key);

            // Attach the original right subtree to the new root of the left subtree.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/075_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Python class that implements a dictionary-like
    set for integers using a Splay Tree.

    This data structure provides amortized O(log n) time complexity for insert,
    delete, and search operations. The key feature of a splay tree is that
    recently accessed elements are moved to the root of the tree, which can
    improve performance for certain access patterns (e.g., temporal locality).
    """

    # Nested class for tree nodes
    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty SplayTree."""
        self.root = None

    # --- Private Helper Methods ---

    def _left_rotate(self, x):
        """Performs a left rotation on the subtree rooted at node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on the subtree rooted at node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x
        
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag case
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case (x is left child of right child)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    # --- Public API Methods ---

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key already exists, the node with that key is splayed to the root.
        If the key does not exist, a new node is created, inserted, and then
        splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key already exists, splay it and return
                self._splay(node)
                return

        # Key does not exist, create and insert new node
        new_node = self._Node(key)
        new_node.parent = parent

        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the newly inserted node to the root
        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the tree.

        This method performs the splaying operation. If the key is found, the
        node containing the key is moved to the root. If the key is not found,
        the last non-null node accessed during the search is moved to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key found, splay the node
                self._splay(node)
                return True
        
        # Key not found, splay the last accessed node (parent of where key would be)
        if last_node:
            self._splay(last_node)
        return False

    def delete(self, key):
        """
        Deletes a key from the tree.

        The search for the key involves a splay operation. If the key is found,
        it is removed, and its subtrees are merged. The in-order predecessor
        (the maximum element in the left subtree) becomes the new root. If the
        key is not found, the tree is splayed based on the last accessed node
        but no deletion occurs.

        Args:
            key (int): The integer key to delete.
        """
        # Search for the key, which splays the node (or its parent) to the root
        if not self.search(key):
            # Key was not found, search() already splayed the tree appropriately
            return

        # After search(key), if the key was found, it's now at the root.
        # However, search() returns true only if root.key == key.
        # We double-check to be safe, especially for an empty tree.
        if self.root is None or self.root.key != key:
            return
        
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left child, so the right subtree becomes the new tree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        elif not right_subtree:
            # No right child, so the left subtree becomes the new tree
            self.root = left_subtree
            if left_subtree:
                left_subtree.parent = None
        else:
            # Both children exist. We need to merge the two subtrees.
            # 1. Make the left subtree the main tree for a moment.
            # 2. Find the maximum element in the left subtree.
            # 3. Splay this maximum element to the root of the left subtree.
            # 4. Attach the right subtree as the right child of the new root.
            
            # Disconnect left subtree from the old root
            left_subtree.parent = None
            
            # Find the max node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Temporarily set the root to perform the splay on the subtree
            self.root = left_subtree
            self._splay(max_node)
            
            # Now, self.root is the max_node of the old left subtree.
            # It has no right child. We can attach the original right subtree.
            self.root.right = right_subtree
            right_subtree.parent = self.root